webui/react/src/components/ExperimentActionDropdown.tsx

```diff
@@ -157,9 +174,15 @@ const ExperimentActionDropdown: React.FC<Props> = ({
     // TODO show loading indicator when we have a button component that supports it.
   };
 
-  const menuItems = getActionsForExperiment(experiment, dropdownActions, curUser).map((action) => (
-    { danger: action === Action.Delete, key: action, label: action }
-  ));
+  const menuItems = getActionsForExperiment(experiment, dropdownActions, curUser)
+    .filter((action) => {
+      const isPinned = settings.pinned.includes(experiment.id);
+      if ((isPinned && action === Action.Pin) || (!isPinned && action === Action.Unpin)) {
+        return false;
+      } else {
+        return true;
+      }
```

Comment:
perf: combining "pin" and "unpin" into "switch pin" would allow you to get rid of this loop